### PR TITLE
ART-1434: shst matching robustness improvements

### PR DESCRIPTION
**Pull request by edmabon**
[Ticket](https://routereports.atlassian.net/browse/ART-1434)

I was seeing some failed matches, even when I'd expect a good match just from looking at the trace, and by running OSRM matching directly. There were quite strict constraints on post-processing the OSRM output, and so I have loosened those off a little bit to give us more robustness in the output.

I will comment on the specific individual changes.

For testing, you can use this trace and run the matching on this branch vs the main branch to see the difference:

```
docker run -it --rm -v ${PWD}:/work --env SHST_TILE_URL=https://rr-binaries.s3.eu-west-2.amazonaws.com/tiles/ shst-image match /work/test.geojson --tile-source=osm/planet-240104 --search-radius=25 --follow-line-direction
```

```
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {},
      "geometry": {
        "type": "LineString",
        "coordinates": [
          [
            -0.2094627999999993,
            52.5977402
          ],
          [
            -0.20944320000000027,
            52.59764479999999
          ],
          [
            -0.20948560000000072,
            52.598114
          ],
          [
            -0.21030939999999987,
            52.598419199999995
          ],
          [
            -0.21161750000000076,
            52.59839629999999
          ],
          [
            -0.22327850000000107,
            52.6007805
          ],
          [
            -0.22510950000000032,
            52.60230259999999
          ],
          [
            -0.22987429999999948,
            52.60523220000001
          ],
          [
            -0.23804459999999988,
            52.60895539999999
          ],
          [
            -0.24569849999999904,
            52.61297230000001
          ],
          [
            -0.24825579999999914,
            52.614566800000006
          ],
          [
            -0.24752409999999983,
            52.614574399999995
          ],
          [
            -0.24688100000000002,
            52.61516189999999
          ],
          [
            -0.24682960000000137,
            52.6155472
          ],
          [
            -0.24776610000000016,
            52.6158524
          ],
          [
            -0.24778540000000043,
            52.61575319999999
          ]
        ]
      }
    }
  ]
}
```

### CODE REVIEW

File: src/commands/match.ts [hunk 1]
```diff
@@ -872,7 +872,7 @@ async function matchLines(outFile, params, lines, flags) {
       var gisRef:SharedStreetsReference = forwardReference(line);
 			
       matchForward = await matcher.matchGeom(line);
-      if(matchForward && matchForward.score < matcher.searchRadius * 2) {
```
Comment (edmabon) [inline src/commands/match.ts:875]: This check of the `score` is somewhat arbitrary - it is calculated with a combination of the difference between the start and end points with the matched traces and the difference between the original length and the matched length.

From our point of view, I reckon if we successfully got a match that was outputted by OSRM, then we are happy. Dealing with failed matches is a pain, so we don't want to have these arbitrary constraints on the output

File: src/tiles.ts [hunk 2]
```diff
@@ -38,14 +38,17 @@ export async function getTilesForId(id:string) {
...
+        return [];
+    } else {
+      var polyBound = bbox(polygon)
+
+      var nwPoint = destination([polyBound[0],polyBound[1]], buffer, 315, {'units':'meters'});
+      var sePoint = destination([polyBound[2],polyBound[3]], buffer, 135, {'units':'meters'});
+      let bounds = [nwPoint.geometry.coordinates[0], nwPoint.geometry.coordinates[1], sePoint.geometry.coordinates[0], sePoint.geometry.coordinates[1]];
+      
+      return getTileIdsForBounds(bounds, false);
+    }
```
Comment (edmabon) [inline src/tiles.ts:51]: This fixes a bug when matching points, see: https://github.com/sharedstreets/sharedstreets-js/pull/94

File: src/graph.ts [hunk 3]
```diff
@@ -902,8 +902,13 @@ export class Graph {
...
             if(visitedEdgeList.length > 0) {
-                var startPoint = turfHelpers.point(feature.geometry.coordinates[0]);
-                var endPoint = turfHelpers.point(feature.geometry.coordinates[feature.geometry.coordinates.length - 1]);
+                // Use the first non-null point (in the matches), rather than just the first point from the raw trace
+                // This way, we are guaranteed to get a match if the OSRM matching was successful
+                var firstNonNull = 0, lastNonNull = matches['tracepoints'].length - 1
+                while (!matches['tracepoints'][firstNonNull] && firstNonNull++);
+                while (!matches['tracepoints'][lastNonNull] && lastNonNull--);
+                var startPoint = turfHelpers.point(matches['tracepoints'][firstNonNull]['location']);
+                var endPoint = turfHelpers.point(matches['tracepoints'][lastNonNull]['location']);
```
Comment (edmabon) [inline src/graph.ts:911]: The start and end point of the journey was previously being found by "snapping" the raw gps to the matched output from OSRM. However, this was sometimes outside the `searchRadius` if OSRM had failed to match the first couple of points, or there is large GPS noise.

Why match again, when OSRM has already mapped the raw points to the route? Instead now, I just grab the first non-null match for the start point, and the final non-null match for the end point.

File: src/graph.ts [hunk 4]
```diff
@@ -855,55 +855,58 @@ export class Graph {
             if(matches['matchings'] &&  matches['matchings'].length > 0) {
                 
                 var match = matches['matchings'][0];
-                if(0 < match.confidence ) {
```
Comment (edmabon) [inline src/graph.ts:858]: If OSRM has managed to match points to a route, that's good enough for us. Let's just match whatever the confidence.

For our application, I'd much rather an incorrect match than no match at all.

**Comment by mandu** [comment]
Thanks @edmabon for clear comments!

**Review by mandu — APPROVED** [review]

File: src/graph.ts (hunk 4 above)
Comment (edmabon) [reply src/graph.ts:858]: > But I get the point if OSRM is able to give a match then shst should not do further filtering.

Yeah this is maybe a better way of putting it! A match with OSRM should _always_ give a successful output as OSRM is matching to exactly the same network.

**Pull request merged**